firmware/library/L3_HAL/apds_9960.hpp

```diff
@@ -1,20 +1,444 @@
+            points_[index_] = Direction::kDown;
+            index_++;
+        }
+      } else
+      {
+        if ((gfifo_data_[i+2] - gfifo_data_[i+3])
+                  < left_sensitivity_)  // Left
+        {
+            points_[index_] = Direction::kLeft;
+            index_++;
+        } else if ((gfifo_data_[i+2] - gfifo_data_[i+3])
+                    > right_sensitivity_)  // Right
+        {
+            points_[index_] = Direction::kRight;
+            index_++;
+        }
+      }
+      if ((gfifo_data_[i] + gfifo_data_[i+1] + gfifo_data_[i+2]
+                  + gfifo_data_[i+3]) > far_sensitivity_ )  // Near
+      {
+          near_count_++;
+      } else if ((gfifo_data_[i] + gfifo_data_[i+1] + gfifo_data_[i+2]
+                  + gfifo_data_[i+3]) < near_sensitivity_ )  // Far
+      {
+          far_count_++;
+      } else
+      {
+          // do nothing
+      }
+    }
+  }
+  char DecodeGestureData() override
+  {
+    for (int i = 0; i < 8; i++)
+    {
+      switch (points_[i])
+      {
+      case Direction::kUp:
+          if (points_[i+1] == Direction::kDown)
+          {
+              // DEBUG_PRINT("Down Gesture\n");
+              return static_cast<char>(Gesture::kSwipeDOWN);
+          }
+          break;
+      case Direction::kDown:
+          if (points_[i+1] == Direction::kUp)
+          {
+              // DEBUG_PRINT("Up Gesture\n");
+              return static_cast<char>(Gesture::kSwipeUP);
+          }
+          break;
+      case Direction::kLeft:
+          if (points_[i+1] == Direction::kRight)
+          {
+              // DEBUG_PRINT("Right Gesture\n");
+              return static_cast<char>(Gesture::kSwipeRIGHT);
+          }
+          break;
+      case Direction::kRight:
+          if (points_[i+1] == Direction::kLeft)
+          {
+              // DEBUG_PRINT("Left Gesture\n");
+              return static_cast<char>(Gesture::kSwipeLEFT);
+          }
+          break;
+      default:
+          if (near_count_ > 5)
+          {
+              // DEBUG_PRINT("Near Gesture\n");
+              return static_cast<char>(Gesture::kNEAR);
+          } else if (far_count_ > 5)
+          {
+              // DEBUG_PRINT("Far Gesture\n");
+              return static_cast<char>(Gesture::kFAR);
+          } else
+          {
+            // do nothing
+          }
+          break;
+      }
+    }
+    return static_cast<char>(Gesture::kERROR);
+  }
+  char GetGesture() override
+  {
+    if (CheckIfGestureOccured())
+    {
+    // check if gesture ended
+      if (ReadGestureMode() == 0)
+      {
+        uint8_t level = 0;
+        // print out raw gesture data
+        DEBUG_PRINT("start gesture:\n");
+        DEBUG_PRINT("flvl: %i\n", GetGestureFIFOLvl());
+        level = GetGestureFIFOLvl();
+
+        ReadGestureFIFO(gfifo_data_, (4*level));
+
+        for (int i = 0; i < (4*level); i+=4)
+        {
+          DEBUG_PRINT("%i, %i, %i, %i\n", gfifo_data_[i], gfifo_data_[i+1] ,
```

Comment:
This would be better suited for your Gesture sensor example, then in your driver.